File: 2016/day3.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

class day3 {
    private static boolean isValidTriangle(Integer[] sides) {
        return sides[0] + sides[1] > sides[2] && sides[0] + sides[2] > sides[1] && sides[1] + sides[2] > sides[0];
    }

    private static long part1(Integer[][] lines) {
        return Arrays.stream(lines).filter(day3::isValidTriangle).count();
    }

    private static long part2(Integer[][] lines) {
        int cnt = 0;

        for (int i = 0; i < lines.length; i += 3)
            for (int j = 0; j < 3; ++j)
                if (isValidTriangle(new Integer[]{lines[i][j], lines[i + 1][j], lines[i + 2][j]}))
                    cnt += 1;

        return cnt;
    }

    public static void main(String[] args) {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        var lines = reader.lines()
                .map(str -> str.trim().split("\\s+"))
                .map(arr -> Arrays.stream(arr).map(Integer::parseInt).toArray(Integer[]::new))
                .toArray(Integer[][]::new);

        System.out.println(part1(lines));
        System.out.println(part2(lines));
    }
}

File: 2015/day3.kt
import java.io.BufferedReader
import java.io.InputStreamReader

private fun part1(line: String): Int {
    var pos = Pair(0, 0)

    val visited = HashSet<Pair<Int, Int>>()
    visited.add(pos)

    for (dir in line) {
        var (x, y) = pos

        when (dir) {
            '>' -> ++x
            '<' -> --x
            'v' -> ++y
            '^' -> --y
        }

        pos = Pair(x, y)
        visited.add(pos)
    }

    return visited.size
}

private fun part2(line: String): Int {
    val pos = mutableListOf(Pair(0, 0), Pair(0, 0))
    var i = 0

    val visited = HashSet<Pair<Int, Int>>()
    visited.add(pos[i])

    for (dir in line) {
        var (x, y) = pos[i]

        when (dir) {
            '>' -> ++x
            '<' -> --x
            'v' -> ++y
            '^' -> --y
        }

        pos[i] = Pair(x, y)
        visited.add(pos[i])
        i = (i + 1) % pos.size
    }

    return visited.size
}

fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val line = reader.readLine()

    println(part1(line))
    println(part2(line))
}
